fancyerror.go

```diff
@@ -0,0 +1,260 @@
+// holds a copy of req. If bodyData is non-nil, it
+// will be used as the data in the Request.Body field;
+// otherwise body data will be read from req.Body.
+func newDecodeRequestError(req *http.Request, bodyData []byte, err error) *DecodeRequestError {
+	if bodyData == nil {
+		bodyData = readBodyForError(req.Body)
+	}
+	req1 := *req
+	req1.Body = ioutil.NopCloser(bytes.NewReader(bodyData))
+
+	return &DecodeRequestError{
+		Request:     &req1,
+		DecodeError: errgo.Mask(err, errgo.Any),
+	}
+}
+
+// DecodeRequestError represents an error when an HTTP
+// request could not be decoded.
+type DecodeRequestError struct {
+	// Request holds the problematic HTTP request.
+	// The body of this does not need to be closed
+	// and may be truncated if the response is large.
+	Request *http.Request
+
+	// DecodeError holds the error that was encountered
+	// when decoding.
+	DecodeError error
+}
+
+func (e *DecodeRequestError) Error() string {
+	return e.DecodeError.Error()
+}
+
+// fancyDecodeError is an error type that tries to
+// produce a nice error message when the content
+// type of a request or response is wrong.
+type fancyDecodeError struct {
+	// contentType holds the contentType of the request or response.
+	contentType string
+
+	// body holds up to maxErrorBodySize saved bytes of the
+	// request or response body.
+	body []byte
+}
+
+func newFancyDecodeError(h http.Header, body io.Reader) *fancyDecodeError {
+	return &fancyDecodeError{
+		contentType: h.Get("Content-Type"),
+		body:        readBodyForError(body),
+	}
+}
+
+func readBodyForError(r io.Reader) []byte {
+	data, _ := ioutil.ReadAll(io.LimitReader(noErrorReader{r}, int64(maxErrorBodySize)))
+	return data
+}
+
+// maxErrorBodySize holds the maximum amount of body that
+// we try to read for an error before extracting text from it.
+// It's reasonably large because:
+// a) HTML often has large embedded scripts which we want
+// to skip and
+// b) it should be an relatively unusual case so the size
+// shouldn't harm.
+//
+// It's defined as a variable so that it can be redefined in tests.
+var maxErrorBodySize = 200 * 1024
+
+// isJSONMediaType reports whether the content type of the given header implies
+// that the content is JSON.
+func isJSONMediaType(header http.Header) bool {
+	contentType := header.Get("Content-Type")
+	mediaType, _, _ := mime.ParseMediaType(contentType)
+	return mediaType == "application/json"
+}
+
+// Error implements error.Error by trying to produce a decent
+// error message derived from the body content.
+func (e *fancyDecodeError) Error() string {
+	mediaType, _, err := mime.ParseMediaType(e.contentType)
+	if err != nil {
+		// Even if there's no media type, we want to see something useful.
+		mediaType = fmt.Sprintf("%q", e.contentType)
+	}
+
+	// TODO use charset.NewReader to convert from non-utf8 content?
+	switch mediaType {
+	case "text/html":
+		text, err := htmlToText(bytes.NewReader(e.body))
```

Comment:
As the comment above says, this can never happen anyway, so I'm worrying myself too much.